libraries/standard/mqtt/src/mqtt.c

```diff
@@ -99,6 +101,509 @@ static int32_t sendPacket( MQTTContext_t * pContext,
+            packetTypeByte = MQTT_PACKET_TYPE_PUBCOMP;
+            packetType = MQTTPubcomp;
+            break;
+        default:
+            /* Nothing to send. */
+            break;
+    }
+
+    if( packetTypeByte != 0U )
+    {
+        status = MQTT_SerializeAck( &( pContext->networkBuffer ),
+                                    packetTypeByte,
+                                    packetId );
+        if( status == MQTTSuccess )
+        {
+            bytesSent = sendPacket( pContext,
+                                    pContext->networkBuffer.pBuffer,
+                                    MQTT_PUBLISH_ACK_PACKET_SIZE );
+        }
+
+        if( bytesSent == ( int32_t ) MQTT_PUBLISH_ACK_PACKET_SIZE )
+        {
+            pContext->controlPacketSent = true;
+            newState = MQTT_UpdateStateAck( pContext,
+                                            packetId,
+                                            packetType,
+                                            MQTT_SEND );
+
+            if( newState == MQTTStateNull )
+            {
+                status = MQTTIllegalState;
+            }
+        }
+        else
+        {
+            IotLogErrorWithArgs( "Failed to send ACK packet: SentBytes=%d, "
+                                 "PacketSize=%u",
+                                 bytesSent,
+                                 MQTT_PUBLISH_ACK_PACKET_SIZE );
+            status = MQTTSendFailed;
+        }
+    }
+
+    if( ( status == MQTTSuccess ) && ( newState != MQTTStateNull ) )
+    {
+        *pPublishState = newState;
+    }
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Send a keep alive PINGREQ if the keep alive interval has elapsed.
+ *
+ * @param[in] pContext Initialized MQTT Context.
+ *
+ * @return #MQTTKeepAliveTimeout if a PINGRESP is not received in time,
+ * #MQTTSendFailed if the PINGREQ cannot be sent, or #MQTTSuccess.
+ */
+static MQTTStatus_t handleKeepAlive( MQTTContext_t * const pContext )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    uint32_t now = 0U, keepAliveMs = 0U;
+
+    assert( pContext != NULL );
+    now = pContext->callbacks.getTime();
+    keepAliveMs = 1000U * ( uint32_t ) pContext->keepAliveIntervalSec;
+
+    /* If keep alive interval is 0, it is disabled. */
+    if( ( keepAliveMs != 0U ) && 
+        ( calculateElapsedTime( now, pContext->lastPacketTime ) > keepAliveMs ) )
+    {
+        if( pContext->waitingForPingResp )
+        {
+            /* Has time expired? */
+            if( calculateElapsedTime( now, pContext->pingReqSendTimeMs ) > 
+                pContext->pingRespTimeoutMs )
+            {
+                status = MQTTKeepAliveTimeout;
+            }
+        }
+        else
+        {
+            status = MQTT_Ping( pContext );
+            /* Sending the packet may have taken some time. Recalculate. */
+            pContext->pingReqSendTimeMs = pContext->callbacks.getTime();
```

Comment:
Minor: Can't we use lastPacketTime here instead of calling getTime() again?

Reply:
Yes, since we know the last packet sent was a PINGREQ. Also, would it make more sense to move these lines to `MQTT_Ping()`?

Reply:
Yes. That will be a good change.

Reply:
Moved